### DISCUSSION TITLE
Dynamic Persistence Database

### DISCUSSION THREAD

**Discussion by ljcbn** [elsa-workflows/elsa-core#5020] [Q&A]
Hello,

I'm trying to implement persistance with Elsa and MongoDb and I need to have variable custom data provider (we are using multitenancy architecture) according the origin of the workflow's request. Do you know how i can procceed to dynamicly change the destination persistance's db ?

Thanks for your help

**Comment by glime-ai** [comment]
To dynamically change the destination persistence database in a multi-tenancy architecture with Elsa and MongoDB, you would typically implement a custom `IWorkflowInstanceStore` and `IWorkflowDefinitionStore` that are aware of your multi-tenancy setup. These custom stores would then resolve the correct MongoDB database connection string based on the tenant information associated with the workflow request.

Here's a high-level approach to achieve this:

1. **Tenant Resolution**: Implement a service that can determine the tenant context from the incoming workflow request. This could be based on a subdomain, a specific header, or any other part of the request that can be used to identify the tenant.

2. **Custom Store Implementation**: Create custom implementations of `IWorkflowInstanceStore` and `IWorkflowDefinitionStore` that use the tenant resolution service to determine the correct MongoDB database for the current tenant.

3. **Dependency Injection**: Register your custom stores in the dependency injection container, ensuring that they are used instead of the default stores provided by Elsa.

4. **Database Connection Management**: In your custom store implementations, manage the MongoDB connections dynamically, creating and disposing of them as needed based on the tenant context.

5. **Workflow Execution**: When a workflow is executed, the custom stores will be invoked, and they will interact with the appropriate MongoDB database for the tenant.

Please note that this is a simplified explanation and the actual implementation can be quite complex, depending on the specifics of your multi-tenancy architecture and requirements.

As of my knowledge cutoff date in April 2023, there is no out-of-the-box support for multi-tenancy in Elsa, so you would need to implement this functionality yourself. Additionally, you should consult the Elsa documentation and source code for the latest information and guidance on implementing custom stores and handling multi-tenancy.

If you need further assistance with the implementation, you may consider reaching out to the Elsa community on GitHub or Stack Overflow, or even seeking professional consultation as mentioned in one of the GitHub issues you referenced.
